pkg/controller/deployment/progress_test.go

```diff
@@ -161,3 +172,162 @@ func TestRequeueStuckDeployment(t *testing.T) {
+		},
+		{
+			name:            "DeploymentProgressing: create Progressing condition if it does not exist",
+			d:               currentDeployment(&pds, 3, 2, 2, 2, []extensions.DeploymentCondition{}),
+			allRSs:          []*extensions.ReplicaSet{newRSWithAvailable("bar", 0, 1, 1)},
+			newRS:           newRSWithAvailable("foo", 3, 2, 2),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionTrue,
+			conditionReason: util.ReplicaSetUpdatedReason,
+		},
+		{
+			name:            "DeploymentComplete: dont update lastTransitionTime if deployment already has Progressing=True",
+			d:               currentDeployment(&pds, 3, 3, 3, 3, []extensions.DeploymentCondition{replicaSetUpdated}),
+			allRSs:          []*extensions.ReplicaSet{},
+			newRS:           newRSWithAvailable("foo", 3, 3, 3),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionTrue,
+			conditionReason: util.NewRSAvailableReason,
+			lastTransition:  testTime,
+		},
+		{
+			name:            "DeploymentComplete: update everything if deployment has Progressing=False",
+			d:               currentDeployment(&pds, 3, 3, 3, 3, []extensions.DeploymentCondition{failedTimedOut}),
+			allRSs:          []*extensions.ReplicaSet{},
+			newRS:           newRSWithAvailable("foo", 3, 3, 3),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionTrue,
+			conditionReason: util.NewRSAvailableReason,
+		},
+		{
+			name:            "DeploymentComplete: create Progressing condition if it does not exist",
+			d:               currentDeployment(&pds, 3, 3, 3, 3, []extensions.DeploymentCondition{}),
+			allRSs:          []*extensions.ReplicaSet{},
+			newRS:           newRSWithAvailable("foo", 3, 3, 3),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionTrue,
+			conditionReason: util.NewRSAvailableReason,
+		},
+		{
+			name:            "DeploymentTimedOut: update status if rollout exceeds Progress Deadline",
+			d:               currentDeployment(&pds, 3, 2, 2, 2, []extensions.DeploymentCondition{replicaSetUpdated}),
+			allRSs:          []*extensions.ReplicaSet{},
+			newRS:           newRSWithAvailable("foo", 3, 2, 2),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionFalse,
+			conditionReason: util.TimedOutReason,
+		},
+		{
+			name:            "DeploymentTimedOut: do not update status if deployment has existing timedOut condition",
+			d:               currentDeployment(&pds, 3, 2, 2, 2, []extensions.DeploymentCondition{failedTimedOut}),
+			allRSs:          []*extensions.ReplicaSet{},
+			newRS:           newRSWithAvailable("foo", 3, 2, 2),
+			conditionType:   extensions.DeploymentProgressing,
+			conditionStatus: v1.ConditionFalse,
+			conditionReason: util.TimedOutReason,
+			lastUpdate:      testTime,
+			lastTransition:  testTime,
+		},
+	}
+
+	fake := fake.Clientset{}
```

Comment:
Move this inside the test loop since it's mutated every time an API action happens.